src/module_4/module_4_model-fitting.md

```diff
@@ -0,0 +1,1214 @@
+```
+
+
+    
+![png](module_4_model-fitting_files/module_4_model-fitting_10_0.png)
+    
+
+
+### Best Linear models result
+
+Para nuestra segunda evaluación, emplearemos el modelo lineal que mostró mejores resultados en el ejercicio anterior: regresión Lasso sin variables categóricas, eliminando variables con una importancia menor a 0.02.
+
+
+```python
+train_cols = ["days_since_purchase_variant_id", "discount_pct", "std_days_to_buy_product_type", "count_pets", "set_as_regular", "avg_days_to_buy_product_type", "days_since_purchase_product_type", "user_order_seq", "avg_days_to_buy_variant_id", "abandoned_before", "global_popularity", "normalised_price", "ordered_before",]
+
+regularization_values = [1e-4, 1e-3, 1e-2, 1e-1, 1, 10]
+
+best_auc = 0
+best_c = None
+best_model = None
+
+for c in regularization_values:
+    lr = make_pipeline(
+        StandardScaler(),
+        LogisticRegression(penalty='l1',C=c, solver='liblinear')
+    )
+    lr.fit(X_train[train_cols], y_train)
+
+    y_proba = lr.predict_proba(X_test[train_cols])[:, 1]
+    
+    roc_auc = roc_auc_score(y_test, y_proba)
+    
+    if roc_auc > best_auc:
+        best_auc = roc_auc
+        best_c = c
+        best_model = lr
+
+y_pred = best_model.predict_proba(X_test[train_cols])[:, 1]
+
+plot_metrics(f"LR: C={best_c}", y_pred, y_test)
+
+y_pred = best_model.predict_proba(X_val[train_cols])[:, 1]
+
+plot_metrics(f"LR: C={best_c}", y_pred, y_val)
+```
+
+
+    
+![png](module_4_model-fitting_files/module_4_model-fitting_12_0.png)
+    
+
+
+    Precision Recall AUC: 0.14909622626798194
+    Roc Curve AUC: 0.8321347472511265
+
+
+
+    
+![png](module_4_model-fitting_files/module_4_model-fitting_12_2.png)
+    
+
+
+    Precision Recall AUC: 0.14868719039041456
+    Roc Curve AUC: 0.8348974327812615
+
+
+## Non Linear Models
+
+El propósito de este ejercicio es superar los resultados que logramos con nuestro mejor modelo lineal de la semana pasada, utilizando modelos no lineales. Con estos modelos no lineales, pretendemos aumentar la complejidad de nuestro modelo con el objetivo de mejorar su rendimiento.
+
+Para empezar, desarrollaremos dos modelos:  Random Forest y Gradient Boosting Trees
+
+## Random Forest
+
+Comenzaremos iterando los diferentes parámetros de nuestro modelo para determinar los valores óptimos para nuestro conjunto de datos.
+
+
+```python
+train_cols = numerical_cols + binary_cols
+
+n_trees_grid = [5, 25, 50, 100]
+
+metrics = {
+    'n_trees': [],
+    'train_auc': [],
+    'val_auc': [],
+    'train_ap': [],
+    'val_ap': [],
+    'train_ce': [],
+    'val_ce': []
+}
+
+for n_trees in n_trees_grid:
+    rf = RandomForestClassifier(n_estimators=n_trees)
+    rf.fit(X_train[train_cols], y_train)
+    
+    train_probs = rf.predict_proba(X_train[train_cols])[:, 1]
+    val_probs = rf.predict_proba(X_val[train_cols])[:, 1]
+    
+    metrics['n_trees'].append(n_trees)
```

Comment:
you are doing something quite similar for every model, which in pseudocode can be summarized as: 
```
for hyperparam in hyperparam_set:
 model[hyperparam].train(X)
 pred_test = model[hyperparam].pred(X_test)
 metrics_train = metr_fnc(pred,y_test)
 pred_train = model[hyperparam].pred(X_train)
 metrics_train = metr_fnc(pred,y_train)
```
can you try to conceptualise this and put it on a function?

Reply:
hey Guille! Totally agree. A conceptualised function would look like:

Function:
def evaluate_model(model, X_train, y_train, X_test, y_test, metrics_function):
    model.fit(X_train, y_train)
    pred_test = model.predict(X_test)
    metrics_test = metrics_function(pred_test, y_test)
    pred_train = model.predict(X_train)
    metrics_train = metrics_function(pred_train, y_train)
    return metrics_test, metrics_train

Dictionary with the hyperparameters we want to test:
hyperparam_set = {...}
for hyperparam in hyperparam_set:
    model = MyModel(hyperparam)
    metrics_test, metrics_train = evaluate_model(
    model, X_train, y_train, X_test, y_test, some_metrics_function
    )